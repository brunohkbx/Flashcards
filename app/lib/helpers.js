import { Platform } from 'react-native';

export const waitFor = ms => {
  return new Promise(resolve => setTimeout(resolve, ms));
};

export const getOS = () => {
  return Platform.OS;
};
