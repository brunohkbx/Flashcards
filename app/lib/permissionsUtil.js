import { Permissions } from 'expo';

export const obtainNotificationPermission = async () => {
  const { status: existingStatus } = await Permissions.getAsync(
    Permissions.NOTIFICATIONS
  );
  let finalStatus = existingStatus;

  if (existingStatus !== 'granted') {
    const { status } = await Permissions.askAsync(Permissions.NOTIFICATIONS);

    finalStatus = status;
  }

  return finalStatus;
};
