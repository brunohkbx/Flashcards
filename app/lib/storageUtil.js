import { AsyncStorage } from 'react-native';
import { pickBy, assign } from 'lodash/object';

export const DECK_STORAGE_KEY       = 'flashcards:deck';
export const SETTINGS_STORAGE_KEY   = 'flashcards:settings';
export const NOTIFICATIONS_KEY      = 'flashcards:notifications';

export const getDecks = () => {
  return AsyncStorage.getItem(DECK_STORAGE_KEY);
}

export const createDeck = deck => {
  return AsyncStorage.mergeItem(DECK_STORAGE_KEY, JSON.stringify(deck));
}

export const deleteDeck = id => {
  return getDecks()
    .then(decks => JSON.parse(decks))
    .then(decks => pickBy(decks, deck => deck.id !== id))
    .then(decks => AsyncStorage.setItem(DECK_STORAGE_KEY, JSON.stringify(decks)));
}

export const editDeck = deck => {
  return getDecks()
    .then(decks => JSON.parse(decks))
    .then(decks => assign(decks, deck))
    .then(decks => AsyncStorage.setItem(
      DECK_STORAGE_KEY, JSON.stringify(decks)
    ));
}

export const loadSettings = () => {
  return AsyncStorage.getItem(SETTINGS_STORAGE_KEY);
}

export const updateSettings = settings => {
  return AsyncStorage.setItem(SETTINGS_STORAGE_KEY, JSON.stringify(settings));
}

export const isScheduledNotification = () => {
  return AsyncStorage.getItem(NOTIFICATIONS_KEY);
}

export const scheduleNotification = () => {
  return AsyncStorage.setItem(NOTIFICATIONS_KEY, JSON.stringify(true));
}

export const unscheduleNotification = () => {
  return AsyncStorage.removeItem(NOTIFICATIONS_KEY);
}
