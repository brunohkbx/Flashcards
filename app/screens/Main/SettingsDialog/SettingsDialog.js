import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { View } from 'react-native';
import {
  Dialog,
  HelperText,
  Paragraph,
  Portal,
  Switch
} from 'react-native-paper';
import { obtainNotificationPermission } from '../../../lib/permissionsUtil';
import { updateSettings } from '../../../actions';

export class SettingsDialog extends Component {
  state = {
    error: false
  }

  onNotificationsChange = async () => {
    const { updateSettings, settings: { receiveNotifications } } = this.props;

    const status = await obtainNotificationPermission();

    if (status !== 'granted')
      return this.setState({ error: true });

    updateSettings({ receiveNotifications: !receiveNotifications })
      .then(() => this.setState(
        {
          enableNotifications: !receiveNotifications,
          error: false
        }
      )
    );
  }

  render() {
    const {
      visible,
      settings: { receiveNotifications },
      onDismiss
    } = this.props;

    const { error } = this.state;

    return (
      <Portal>
        <Dialog visible={visible} onDismiss={onDismiss}>
          <Dialog.Title>Settings</Dialog.Title>
          <Dialog.Content>
            <View style={{ flexDirection: 'row', alignItems: 'center', justifyContent: 'space-between' }}>
              <Paragraph>Notifications</Paragraph>
              <Switch
                value={receiveNotifications}
                onValueChange={this.onNotificationsChange}
              />
            </View>
            {error &&
              <HelperText
                type="error"
                style={{ paddingHorizontal: 0 }}
              >
                You need to grant NOTIFICATIONS permission
              </HelperText>
            }
          </Dialog.Content>
        </Dialog>
      </Portal>
    );
  }
}

SettingsDialog.propTypes = {
  visible: PropTypes.bool,
  onDismiss: PropTypes.func.isRequired
};

SettingsDialog.defaultProps = {
  visible: false
};

const mapStateToProps = ({ settings }) => ({ settings });

const mapDispatchToProps = dispatch => ({
  updateSettings: settings => dispatch(updateSettings(settings)),
});

export default connect(mapStateToProps, mapDispatchToProps)(SettingsDialog);
