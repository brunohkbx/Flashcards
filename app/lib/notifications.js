import { Notifications } from 'expo';
import * as storageUtil from '../lib/storageUtil';

export const clearLocalNotification = () => {
  return storageUtil.unscheduleNotification()
    .then(Notifications.cancelAllScheduledNotificationsAsync);
};

export const scheduleLocalNotification = () => {
  return storageUtil.isScheduledNotification()
    .then(JSON.parse)
    .then(data => {
      if (data === null) {
        Notifications.cancelAllScheduledNotificationsAsync();

        let tomorrow = new Date();
        tomorrow.setDate(tomorrow.getDate() + 1);
        tomorrow.setHours(18);
        tomorrow.setMinutes(30);

        Notifications.scheduleLocalNotificationAsync(
          {
            title: 'Haven\'t seen you in a while',
            body: 'ðŸ‘‹ Don\'t forget to take your quiz today!',
            ios: {
              sound: true
            },
            android: {
              sound: true,
              priority: 'high',
              sticky: false,
              vibrate: true
            }
          },
          {
            time: tomorrow,
            repeat: 'day',
          }
        );

        storageUtil.scheduleNotification();
      }
    });
};
