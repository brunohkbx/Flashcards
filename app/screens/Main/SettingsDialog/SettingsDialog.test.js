import React from 'react';
import { shallow } from 'enzyme';
import { SettingsDialog } from './SettingsDialog';
import * as permissionsUtil from '../../../lib/permissionsUtil';

describe('SettingsDialog', () => {
  const setup = propOverrides => {
    const defaultProps = Object.assign({
      onDismiss: jest.fn(),
      settings: {
        receiveNotifications: false
      }
    }, propOverrides);

    const wrapper = shallow(<SettingsDialog {...defaultProps} />);
    const wrapperInstance = wrapper.instance();

    return { wrapper, wrapperInstance };
  };

  it('renders properly', () => {
    const { wrapper } = setup();

    expect(wrapper).toMatchSnapshot();
  });

  it('renders properly when notification permission is not granted', () => {
    const { wrapper } = setup();

    wrapper.setState({ error: true });

    expect(wrapper).toMatchSnapshot();
  });

  describe('onNotificationsChange', () => {
    it('displays an error message when notification permission is not granted', async () => {
      const { wrapperInstance } = setup();
      jest.spyOn(permissionsUtil, 'obtainNotificationPermission').mockImplementation(() => {
        return false;
      });
      jest.spyOn(wrapperInstance, 'setState');

      await wrapperInstance.onNotificationsChange();

      expect(wrapperInstance.setState).toHaveBeenCalledWith({ error: true });
    });

    it('calls updateSettings and toggle enableNotifications in the store', async () => {
      const mockUpdateSettings = jest.fn(() => Promise.resolve());
      const { wrapperInstance } = setup({ updateSettings: mockUpdateSettings });
      jest.spyOn(permissionsUtil, 'obtainNotificationPermission').mockImplementation(() => {
        return 'granted';
      });
      jest.spyOn(wrapperInstance, 'setState');

      await wrapperInstance.onNotificationsChange();

      expect(mockUpdateSettings).toHaveBeenCalledWith(
        { receiveNotifications: !wrapperInstance.props.receiveNotifications }
      );
      expect(wrapperInstance.setState).toHaveBeenCalledWith(
        {
          enableNotifications: !wrapperInstance.props.receiveNotifications,
          error: false
        }
      );
    });
  });
});
