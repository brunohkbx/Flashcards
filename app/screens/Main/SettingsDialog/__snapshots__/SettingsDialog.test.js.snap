// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SettingsDialog renders properly 1`] = `
<withTheme(Portal)>
  <withTheme(Dialog)
    onDismiss={[MockFunction]}
    visible={false}
  >
    <withTheme(Dialog.Title)>
      Settings
    </withTheme(Dialog.Title)>
    <Dialog.Content>
      <Component
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "space-between",
          }
        }
      >
        <Paragraph>
          Notifications
        </Paragraph>
        <withTheme(Switch)
          onValueChange={[Function]}
          value={false}
        />
      </Component>
    </Dialog.Content>
  </withTheme(Dialog)>
</withTheme(Portal)>
`;

exports[`SettingsDialog renders properly when notification permission is not granted 1`] = `
<withTheme(Portal)>
  <withTheme(Dialog)
    onDismiss={[MockFunction]}
    visible={false}
  >
    <withTheme(Dialog.Title)>
      Settings
    </withTheme(Dialog.Title)>
    <Dialog.Content>
      <Component
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "space-between",
          }
        }
      >
        <Paragraph>
          Notifications
        </Paragraph>
        <withTheme(Switch)
          onValueChange={[Function]}
          value={false}
        />
      </Component>
      <withTheme(HelperText)
        style={
          Object {
            "paddingHorizontal": 0,
          }
        }
        type="error"
      >
        You need to grant NOTIFICATIONS permission
      </withTheme(HelperText)>
    </Dialog.Content>
  </withTheme(Dialog)>
</withTheme(Portal)>
`;
